spring/spring4/spring4-common/src/main/java/org/infinispan/spring/session/AbstractInfinispanSessionRepository.java

```diff
@@ -0,0 +1,90 @@
+package org.infinispan.spring.session;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+
+import org.infinispan.spring.provider.SpringCache;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.context.ApplicationEventPublisherAware;
+import org.springframework.session.MapSession;
+import org.springframework.session.SessionRepository;
+
+/**
+ * Infinispan implementation for Spring Session with basic functionality.
+ *
+ * @author Sebastian Łaskawiec
+ * @see <a href="http://projects.spring.io/spring-session">Spring Session Web Page</a>
+ * @see SessionRepository
+ * @see ApplicationEventPublisherAware
+ * @since 9.0
+ */
+public abstract class AbstractInfinispanSessionRepository implements SessionRepository<MapSession>, ApplicationEventPublisherAware, InitializingBean {
+
+   protected final AbstractApplicationPublisherBridge applicationEventPublisher;
+   protected final SpringCache cache;
+   protected final PrincipalNameResolver principalNameResolver = new PrincipalNameResolver();
+
+   protected AbstractInfinispanSessionRepository(SpringCache cache, AbstractApplicationPublisherBridge eventsBridge) {
+      Objects.requireNonNull(cache, "SpringCache can not be null");
+      Objects.requireNonNull(eventsBridge, "EventBridge can not be null");
+      applicationEventPublisher = eventsBridge;
+      this.cache = cache;
+   }
+
+   @Override
+   public void afterPropertiesSet() throws Exception {
+      applicationEventPublisher.registerListener();
```

Comment:
When will the listener be unregistered? Maybe detect when the ctx is being shutdown and clean it up?